### ISSUE TITLE
Binary-only streaming API to the Chain DB

### ISSUE THREAD

**Issue by edsko** [input-output-hk/ouroboros-network#1178]
The chain DB storage layer (specifically, the immutable DB) was very specifically designed so that we could stream blocks directly from disk without having to do _any_ interpretation: the file is literally just the concatenation of the blocks. However, we don't currently take advantage of this in the block fetch server, which deserialises and serialises all requested blocks. People have already complained about CPU usage of the node when serving blocks, so it is important that we address this.

A complication here is that we need to check hashes in order to verify the bounds of the requested blocks. However, if block fetch only requests small ranges, then that hash check will be expensive. Not entirely clear to me at this stage what the right solution to that problem is.

**Comment by mrBliss** [comment]
> A complication here is that we need to check hashes in order to verify the bounds of the requested blocks. However, if block fetch only requests small ranges, then that hash check will be expensive. Not entirely clear to me at this stage what the right solution to that problem is.

These checks are only done when streaming blocks from the ImmutableDB, because the ImmutableDB is indexed by slots only. These checks are done before opening the ImmutableDB iterator so that we know we'll be able to stream the right blocks.

But, see the docstring of [`ChainDB.streamBlocks`](
https://github.com/input-output-hk/ouroboros-network/blob/2441f25ee5a2da0fbf008f3e9a1b1fd4e57f1de3/ouroboros-consensus/src/Ouroboros/Storage/ChainDB/API.hs#L186):
```haskell
-- Streaming blocks older than @k@ is permitted, but only when they are
-- part of the current fork (at the time of initialization). Streaming a
-- fork that forks off more than @k@ blocks in the past is not permitted
-- and an 'UnknownRange' error will be returned in that case.
```
If we make a (hard) precondition of the first sentence, then we don't need to do this check at all. In other words, if a peer requests blocks older than `k`, we simply send them the blocks from our ImmutableDB, regardless the hashes of the requested blocks. The peer's Block Fetch client will check the hashes anyway using `blockMatchesHeader` and will abort (?) after the first block. As Edsko puts it: garbage in, garbage out :slightly_smiling_face:

(We should still keep the `UnknownRange` error, because the requested range can still lie in the VolatileDB, in which case it is cheap to check the hashes, as they are stored in an in-memory index.)

Note that this can happen with non-adversarial clients too:
> A client is at around `k` back in following our chain, we get a near-`k` long fork, and the client ends up waiting several slots between the chain sync and the block fetch, and now asks for blocks covering a span that is now immutable for us, but forked between the last chain sync message and the block fetch.

**Comment by edsko** [comment]
Conclusion: we might end up serving the wrong blocks to non-adversarial clients if we switch fork after they send requests for blocks, and the slot numbers they ask for have become immutable but with different blocks. Unless we implement a hash check, we don't notice, so that we will serve the wrong blocks; the block fetch client will notice this ("does the block match the header") and disconnect from us, potentially classifying us as adversarial. It's not _terribly_ likely to happen; it will only happen if we either switch to fork which forks off way back, or we switch to a very long new fork (so that even blocks near our old tip are suddenly immutable). 

It's not a huge disaster if it happens, but the better fix will be to (1) modify the index files to have hashes for _all_ blocks and (2) keep a global cache of index files in the imm DB (with some kind of simple cache replacement scheme) so that requests for consecutive ranges of the imm DB (a very likely scenario) won't involve re-reading an index file each time.

**Comment by edsko** [comment]
Since (1) changes the binary format, we should do it ASAP.

**Comment by mrBliss** [comment]
Partially addressed by #1271. The remaining work is moved to #1330.
